### ISSUE TITLE
Inconsistent behaviour between `form(action=...)` and `input(type='submit', formaction=...)`

### ISSUE THREAD

**Issue by MiguelGuthridge** [COMP1010UNSW/pyhtml-enhanced#16]
# What's the issue

Currently, `pyhtml-enhanced` mimics the behaviour of `pyhtml` for `form` and `input(type='submit')` elements, in the sense that

* `form` elements are given the default attribute `method="POST"`
* `input(type='submit')` elements are not given a `formmethod` attribute

[meta/tags.yml:38](https://github.com/COMP1010UNSW/pyhtml-enhanced/blob/baf0cf94895179877cf72c2e7e47f28e6a8b3920/meta/tags.yml#L38)
```yml
form:
  attributes:
    method:
      doc: The HTTP request method to use when submitting this form. In almost all cases, you'll want this to be POST.
      default: "'POST'"
    action: The URL to request to when submitting this form. By default, requests will be sent to the same URL as the current page.
```

This means that by default, submitting a form will send a `POST` request, unless the `input`'s `formaction` is specified, in which case, a `GET` request is sent (since `formmethod` defaults to `GET` when unspecified), which is very confusing.

```py
import pyhtml as p

# Action is specified in `<form>` element
print(str(p.form(action='/test')(
    p.input(placeholder='Enter some text'),
    # Clicking `Submit` will send a `POST` request
    p.input(type='submit'),
)))
```

```py
import pyhtml as p

# Action is specified in `<input>` element
print(str(p.form(
    p.input(placeholder='Enter some text'),
    # Clicking `Submit` will send a `GET` request
    p.input(type='submit', formaction='/test'),
)))
```

# How to fix it

To address this, there are a few possible solutions:

## 1. Stop specifying the `action` attribute on `form` elements by default

This is a relatively simple fix to implement, and means that `pyhtml-enhanced` will behave similarly to regular HTML. However, in almost all cases, users will want to make their `form` send a post request, and `p.form(method='POST')(...)` is much more tedious to write than `p.form(...)`. Additionally, this is a breaking change, and would require most users to update their code to update almost all instances of `form` elements, or they will start sending `GET` requests.

## 2. Make the `formmethod` attribute be specified for all `input(type='submit')` elements where `formaction` is also specified

This would require a bunch of additional complexity, since whether to specify the `formmethod` property would depend on the value of the `type` attribute and the presence of the `formaction` attribute, which currently isn't possible due to the design of the code generation system. We would need to write support for this, or implement the `input` element manually. implementing `input` manually would also have the benefit of allowing us to provide smarter inline suggestions by using the `@typing.override` decorator on the `__init__` method.

Additionally, this would make `pyhtml-enhanced` more detached from the actual behaviour of HTML, which somewhat defeats its goal of being the Python equivalent of writing regular HTML. However, I will note that as with the original `pyhtml` library, some default attributes are specified for other elements (eg `script` and `style`) to aid in simplicity.

Like option 1, this is also a breaking change, however the scope of the damage to dependent projects is likely to be significantly lower, since far fewer rely on this behaviour, although it is enough that I would still be concerned about the compatibility.

If this fix was chosen, this behaviour would need to be documented, including information on how to "escape" it (by manually specifying `formmethod`, similar to specifying `method` on `form` elements).

## 3. Don't fix it - just document the behaviour

This is the easiest thing to implement, and results in no breaking changes. It means that `pyhtml-enhanced` will remain a drop-in replacement for `pyhtml`, which is a nice benefit. However, it means that people will still remain confused about the behaviour the first time they encounter it.

# Considerations when deciding on a solution

This library is primarily intended to be used for educational purposes, and small inconsistencies can be immensely confusing for students with only a few weeks of programming under their belts. As such, I personally lean towards options 1 and 2 over option 3, since they produce consistent behaviour. When it comes to those two options, I think that option 2 would result in overall nicer code, and the required changes would likely make the library more flexible if similar changes are required in the future. However, I am still a little concerned with how it makes the behaviour different to the HTML that is implied by the code - perhaps that is just my inner computer scientist showing though - I doubt most students would care, and if any did, the added simplicity would be easy to justify.

**Comment by cenkalti** [comment]
Hi @MiguelGuthridge . Option 2 makes more sense to me. If option 1 is chosen to make the library close to HTML as possible, the same idea would also apply to existing `script` and `style` tags which also has default attributes for convenience.

I have implemented this as a PoC. Please take a look: https://github.com/cenkalti/pyhtml/commit/eaa76186088a88e0c1ed347d62955c89d361ff60

**Comment by MiguelGuthridge** [comment]
Fixed in [`3f53f2b`](https://github.com/COMP1010UNSW/pyhtml-enhanced/commit/3f53f2b665ee8f28fb4081f203f7f3c259705585)
